tensorflow/core/util/cuda_kernel_helper.h

```diff
@@ -109,14 +110,52 @@ inline Cuda2DLaunchConfig GetCuda2DLaunchConfig(int xdim, int ydim,
 
   config.thread_per_block = dim3(block_cols, block_rows, 1);
 
-  int grid_x = std::min((xdim + block_cols - 1) / block_cols, max_blocks);
+  int grid_x = std::min(DIV_UP(xdim, block_cols), max_blocks);
 
   config.block_count = dim3(
       grid_x, std::min(max_blocks / grid_x, std::max(ydim / block_rows, 1)), 1);
 
   return config;
 }
 
+// Calculate the Cuda 2D and 3D launch config we should use for a kernel launch.
+// This variant takes the resource limits of func into account to maximize occupancy.
+using Cuda3DLaunchConfig = Cuda2DLaunchConfig;
+
+template <typename DeviceFunc>
+inline Cuda3DLaunchConfig GetCuda3DLaunchConfig(int xdim, int ydim, int zdim,
+                                                const GPUDevice& d, DeviceFunc func,
+                                                size_t dynamic_shared_memory_size = 0) {
+  int block_count = 0;
+  int thread_per_block = 0;
+  cudaOccupancyMaxPotentialBlockSize(&block_count, &thread_per_block, func,
+                                     dynamic_shared_memory_size,
+                                     xdim * ydim * zdim);
+
+  int threadsx = std::min(xdim, thread_per_block);
+  int threadsy = std::min(ydim, std::max(thread_per_block / threadsx, 1));
+  int threadsz = std::min(zdim, std::max(thread_per_block / (threadsx * threadsy), 1));
```

Comment:
This can cause division by zero, e.g., if xdim is zero.  Please add handling for one or mode dims being <= 0.